File: examples/parse_resume.py
#!/usr/bin/env python3
"""Client example: call the Parselex /inference-v2/parse API and print structured JSON.

With no args, parses the bundled demo resume (full-database/Karan.pdf) and saves
the result to examples/output/Karan.json.

Usage:
    python3 parse_resume.py
    python3 parse_resume.py resume.pdf
    python3 parse_resume.py resume.pdf --out result.json
    python3 parse_resume.py resume.pdf --url http://localhost:8000 --precision int8
"""
from __future__ import annotations

import argparse
import json
import os
import sys
import urllib.error
import urllib.request
import uuid

_SCRIPT_DIR = os.path.dirname(os.path.abspath(__file__))
_DEFAULT_PDF = os.path.join(_SCRIPT_DIR, "..", "full-database", "Karan.pdf")


def parse_resume(pdf_path: str, base_url: str, precision: str) -> dict:
    boundary = uuid.uuid4().hex
    with open(pdf_path, "rb") as f:
        pdf_bytes = f.read()

    body = (
        f"--{boundary}\r\n"
        f'Content-Disposition: form-data; name="file"; filename="{pdf_path.split("/")[-1]}"\r\n'
        f"Content-Type: application/pdf\r\n\r\n"
    ).encode() + pdf_bytes + f"\r\n--{boundary}--\r\n".encode()

    url = f"{base_url.rstrip('/')}/inference-v2/parse?precision={precision}"
    req = urllib.request.Request(
        url,
        data=body,
        headers={"Content-Type": f"multipart/form-data; boundary={boundary}"},
        method="POST",
    )
    with urllib.request.urlopen(req, timeout=120) as resp:
        return json.loads(resp.read())


def main() -> None:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("pdf", nargs="?", default=_DEFAULT_PDF, help="Path to resume PDF (default: bundled demo resume)")
    parser.add_argument("--url", default="http://localhost:8000", help="Parselex engine base URL")
    parser.add_argument("--precision", default="fp32", choices=["fp32", "int8"])
    parser.add_argument("--out", help="Write JSON output here (default: <script dir>/output/<pdf name>.json)")
    args = parser.parse_args()

    if not args.out:
        stem = os.path.splitext(os.path.basename(args.pdf))[0]
        args.out = os.path.join(_SCRIPT_DIR, "output", f"{stem}.json")

    try:
        result = parse_resume(args.pdf, args.url, args.precision)
    except urllib.error.HTTPError as exc:
        print(f"HTTP {exc.code}: {exc.read().decode()}", file=sys.stderr)
        sys.exit(1)
    except urllib.error.URLError as exc:
        print(f"Could not reach {args.url} — is the engine running? ({exc.reason})", file=sys.stderr)
        sys.exit(1)

    output = json.dumps(result["structured"], indent=2, ensure_ascii=False)
    if args.out:
        os.makedirs(os.path.dirname(args.out), exist_ok=True)
        with open(args.out, "w") as f:
            f.write(output)
        print(f"Wrote structured JSON to {args.out}")
    else:
        print(output)


if __name__ == "__main__":
    main()

File: examples/parse_resume.mjs
#!/usr/bin/env node
// Client example: call the Parselex /inference-v2/parse API and print structured JSON.
//
// With no args, parses the bundled demo resume (full-database/Karan.pdf) and
// saves the result to examples/output/Karan.json.
//
// Usage:
//   node parse_resume.mjs
//   node parse_resume.mjs resume.pdf
//   node parse_resume.mjs resume.pdf --out result.json
//   node parse_resume.mjs resume.pdf --url http://localhost:8000 --precision int8

import { mkdir, readFile, writeFile } from 'node:fs/promises';
import { basename, dirname, extname, join } from 'node:path';
import { fileURLToPath } from 'node:url';

const SCRIPT_DIR = dirname(fileURLToPath(import.meta.url));
const DEFAULT_PDF = join(SCRIPT_DIR, '..', 'full-database', 'Karan.pdf');

function parseArgs(argv) {
  const args = { pdf: null, url: 'http://localhost:8000', precision: 'fp32', out: null };
  const rest = [];
  for (let i = 0; i < argv.length; i++) {
    if (argv[i] === '--url') args.url = argv[++i];
    else if (argv[i] === '--precision') args.precision = argv[++i];
    else if (argv[i] === '--out') args.out = argv[++i];
    else rest.push(argv[i]);
  }
  args.pdf = rest[0] ?? DEFAULT_PDF;
  if (!args.out) {
    const stem = basename(args.pdf, extname(args.pdf));
    args.out = join(SCRIPT_DIR, 'output', `${stem}.json`);
  }
  return args;
}

async function parseResume(pdfPath, baseUrl, precision) {
  const bytes = await readFile(pdfPath);
  const form = new FormData();
  form.append('file', new Blob([bytes], { type: 'application/pdf' }), basename(pdfPath));

  const url = `${baseUrl.replace(/\/$/, '')}/inference-v2/parse?precision=${precision}`;
  const res = await fetch(url, { method: 'POST', body: form });
  const data = await res.json();
  if (!res.ok) {
    throw new Error(`HTTP ${res.status}: ${JSON.stringify(data)}`);
  }
  return data;
}

async function main() {
  const args = parseArgs(process.argv.slice(2));

  let result;
  try {
    result = await parseResume(args.pdf, args.url, args.precision);
  } catch (err) {
    console.error(`Failed to reach ${args.url} — is the engine running?\n${err.message}`);
    process.exit(1);
  }

  const output = JSON.stringify(result.structured, null, 2);
  await mkdir(dirname(args.out), { recursive: true });
  await writeFile(args.out, output);
  console.log(`Wrote structured JSON to ${args.out}`);
}

main();
